src/gui/src/clockWidget.cpp

```diff
@@ -0,0 +1,1202 @@
+}
+
+void ClockTreeView::selectionChanged()
+{
+  for (const auto& sel : scene_->selectedItems()) {
+    QVariant data = sel->data(0);
+    odb::dbBTerm* bterm = data.value<odb::dbBTerm*>();
+    if (bterm != nullptr) {
+      emit selected(Gui::get()->makeSelected(bterm));
+    } else {
+      odb::dbITerm* iterm = data.value<odb::dbITerm*>();
+      if (iterm != nullptr) {
+        emit selected(Gui::get()->makeSelected(iterm));
+      } else {
+        odb::dbNet* net = data.value<odb::dbNet*>();
+        if (net != nullptr) {
+          emit selected(Gui::get()->makeSelected(net));
+        } else {
+          odb::dbInst* inst = data.value<odb::dbInst*>();
+          if (inst != nullptr) {
+            emit selected(Gui::get()->makeSelected(inst));
+          }
+        }
+      }
+    }
+  }
+}
+
+const char* ClockTreeView::getClockName() const
+{
+  return tree_->getClock()->name();
+}
+
+void ClockTreeView::showRenderer(bool show) const
+{
+  if (renderer_ == nullptr) {
+    return;
+  }
+
+  auto* gui = Gui::get();
+  if (show) {
+    gui->registerRenderer(renderer_.get());
+  } else {
+    gui->unregisterRenderer(renderer_.get());
+  }
+}
+
+void ClockTreeView::enableRenderer(bool enable)
+{
+  if (enable) {
+    renderer_ = std::make_unique<ClockTreeRenderer>(tree_.get());
+    scene_->setRendererEnabled(true);
+    showRenderer(true);
+  } else {
+    scene_->setClearPathEnable(false);
+    renderer_ = nullptr;
+  }
+}
+
+qreal ClockTreeView::convertDelayToY(sta::Delay delay) const
+{
+  return (delay - min_delay_) / (max_delay_ - min_delay_) * time_scale_;
+}
+
+sta::Delay ClockTreeView::convertYToDelay(qreal y) const
+{
+  return y / time_scale_ * (max_delay_ - min_delay_) + min_delay_;
+}
+
+QString ClockTreeView::convertDelayToString(sta::Delay delay) const
+{
+  const sta::Unit* unit = tree_->getNetwork()->units()->timeUnit();
+  std::string sdelay = unit->asString(delay, 3);
+  sdelay += " ";
+  sdelay += unit->scaleAbreviation();
+  sdelay += unit->suffix();
+  return QString::fromStdString(sdelay);
+}
+
+std::vector<ClockNodeGraphicsViewItem*> ClockTreeView::buildTree(
+    const ClockTree* tree,
+    const STAGuiInterface* sta,
+    int center_index)
+{
+  center_index = std::max(
+      std::min(center_index, static_cast<int>(bin_center_.size() - 1)), 0);
+
+  sta::dbNetwork* network = sta->getNetwork();
+
+  std::vector<ClockNodeGraphicsViewItem*> drivers;
+  QPolygonF driver_bbox;
+  qreal driver_offset = 0;
+  ClockTree* parent_tree = tree->getParent();
+  for (const auto& [driver, arrival] : tree->getDriverDelays()) {
+    PinArrival output_pin{driver, arrival};
+
+    ClockNodeGraphicsViewItem* node;
+    if (tree->isRoot()) {
+      node = addRootToScene(driver_offset, output_pin, network);
+    } else {
+      PinArrival input_pin;
+      if (parent_tree != nullptr) {
+        const auto in = parent_tree->getPairedReciever(driver);
+        input_pin.pin = in.first;
+        input_pin.delay = in.second;
```

Comment:
As you used elsewhere:
 const auto& [receiver, time] = parent_->getPairedReciever(driver);